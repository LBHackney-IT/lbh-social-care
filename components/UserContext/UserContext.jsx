import { createContext, useContext } from 'react';
import { useRouter } from 'next/router';

import { shouldRedirect } from 'utils/auth';
import { isBrowser } from 'utils/ssr';

export const UserContext = createContext({
  user: undefined,
  setUser: () => {},
});

export const AuthProvider = ({ children, user }) => {
  const { pathname, push } = useRouter();
  const redirect = isBrowser() && shouldRedirect(pathname, user);
  if (redirect) {
    push(redirect);
    return <></>;
  }
  return (
    <UserContext.Provider
      value={{
        user,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};

export function useAuth() {
  const context = useContext(UserContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
}
