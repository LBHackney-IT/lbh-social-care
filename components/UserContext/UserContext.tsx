import { createContext, useContext, ReactNode, ReactElement } from 'react';
import { useRouter } from 'next/router';

import { shouldRedirect } from 'utils/auth';
import { isBrowser } from 'utils/ssr';

import type { User } from 'types';

interface Props {
  children: ReactNode;
  user?: User;
}

export const UserContext = createContext<{ user?: User }>({
  user: undefined,
});

export const AuthProvider = ({ children, user }: Props): ReactElement => {
  const { pathname, push } = useRouter();
  const redirect = isBrowser() && shouldRedirect(pathname, user);
  if (redirect) {
    push(redirect);
    return <></>;
  }
  return (
    <UserContext.Provider
      value={{
        user,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};

export const useAuth = (): { user?: User } => {
  const context = useContext(UserContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
};
