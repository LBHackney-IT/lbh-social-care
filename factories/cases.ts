import { Factory } from 'fishery';

import { Case } from 'types';

const caseFactory = Factory.define<Case>(({ sequence }) => ({
  recordId: sequence.toString(),
  personId: 123,
  dateOfEvent: '25/10/2020 13:49:43',
  formName: 'foorm',
  officerEmail: 'Fname.Lname@hackney.gov.uk',
  caseFormData: {
    note: 'I am the note',
    first_name: 'Foo',
    last_name: 'Bar',
    timestamp: '11/03/2021 16:51:56',
    mosaic_id: 123,
    form_name_overall: 'ASC_case_note',
    worker_email: 'worker@hackney.gov.uk',
    form_name: 'Foo bar',
    context_flag: 'A',
    date_of_event: '25/10/2020 13:49:43',
  },
}));

const allocationCaseFactory = caseFactory.params({
  formName: 'Worker allocated',
  caseFormUrl: undefined,
  caseFormData: {
    allocation_id: 321,
    created_by: 'foo@hackney.gov.uk',
    form_name_overall: 'API_Allocation',
    form_name: 'Worker allocated',
  },
});

const deallocationCaseFactory = caseFactory.params({
  formName: 'Worker allocated',
  caseFormUrl: undefined,
  caseFormData: {
    allocation_id: 321,
    created_by: 'foo@hackney.gov.uk',
    deallocation_reason: 'deallocated because',
    form_name_overall: 'API_Deallocation',
    form_name: 'Worker deallocated',
  },
});

export const mockedNote = caseFactory.build({ caseFormUrl: 'https://foo.bar' });
export const mockedAllocationNote = allocationCaseFactory.build();
export const mockedDeallocationNote = deallocationCaseFactory.build();
