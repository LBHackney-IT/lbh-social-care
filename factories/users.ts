import { Factory } from 'fishery';

import { User } from 'types';

export const userFactory = Factory.define<User>(() => ({
  name: 'foo',
  hasAdminPermissions: true,
  hasChildrenPermissions: true,
  hasAdultPermissions: true,
  email: 'foo@bar.com',
  permissionFlag: 'A',
  isAuthorised: true,
}));

export const mockedUser = userFactory.build();
