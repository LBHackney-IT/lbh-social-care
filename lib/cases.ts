import axios from 'axios';

import type { Case, CaseData } from 'types';

const { ENDPOINT_API, AWS_KEY } = process.env;

const headersWithKey = {
  'x-api-key': AWS_KEY,
};

const regex = /"_id.*ObjectId\(.*\),./;

const sanitiseCaseFormData = (caseFormData: string): string =>
  caseFormData && JSON.parse(caseFormData.replace(regex, ''));

export const getCases = async (
  params: Record<string, unknown>
): Promise<CaseData> => {
  const { data } = await axios.get(`${ENDPOINT_API}/cases`, {
    headers: headersWithKey,
    params: params,
  });
  return {
    ...data,
    cases: data.cases?.map((c: Record<string, string>) => ({
      ...c,
      caseFormData: sanitiseCaseFormData(c.caseFormData),
    })),
  };
};

export const getCasesByResident = (
  mosaic_id: number,
  params: Record<string, unknown>
): Promise<CaseData> => getCases({ mosaic_id, ...params });

export const getCaseByResident = async (
  mosaic_id: number,
  case_id: string,
  params: Record<string, unknown>
): Promise<Case | undefined> => {
  // TODO: this should be covered by the BE
  const data = await getCases({ mosaic_id, ...params, limit: 100 });
  return data.cases?.find(({ recordId }) => recordId === case_id);
};

export const addCase = async (
  formData: Record<string, unknown>
): Promise<{ ref: string }> => {
  const { data } = await axios.post(`${ENDPOINT_API}/cases`, formData, {
    headers: { 'Content-Type': 'application/json', 'x-api-key': AWS_KEY },
  });
  return data;
};
