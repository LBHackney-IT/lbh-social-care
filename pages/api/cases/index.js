import { postResidentCase } from 'utils/server/residents';
import { isAuthorised } from 'utils/auth';

export default async (req, res) => {
  if (!isAuthorised({ req })) {
    return res.status(401).send('Auth cookie missing.');
  }
  switch (req.method) {
    case 'POST':
      try {
        const { data } = await postResidentCase(req.query.id, req.body);
        res.setHeader('Content-Type', 'application/json');
        res.end(JSON.stringify(data));
      } catch (error) {
        console.log(error.status);
        console.log('Cast post error:', error);
        res.status(500).json('Unable to post cast');
      }
      break;

    default:
      res.status(400).json('Invalid request method');
      console.log(res.status);
  }
};
