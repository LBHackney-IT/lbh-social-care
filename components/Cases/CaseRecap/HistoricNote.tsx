import ErrorMessage from 'components/ErrorMessage/ErrorMessage';
import Spinner from 'components/Spinner/Spinner';
import PageView from 'components/PageView/PageView';
import { useCaseNote } from 'utils/api/cases';

interface Props {
  recordId: string;
}

const HistoricNote = ({ recordId }: Props): React.ReactElement => {
  const { data: record, error: recordError } = useCaseNote(recordId);

  if (recordError) {
    return <ErrorMessage />;
  }
  if (!record) {
    return <Spinner />;
  }
  return (
    <PageView
      title={record.title}
      officerEmail={record.officerEmail}
      formName={record.formName}
      dateOfEvent={record.dateOfEvent}
      content={record.content}
    />
  );
};

export default HistoricNote;
