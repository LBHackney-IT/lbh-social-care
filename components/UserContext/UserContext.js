import { createContext, useContext, useEffect, useState } from 'react';
import { useRouter } from 'next/router';

import { AUTH_WHITELIST } from 'utils/auth';
import { getUser } from 'utils/api/checkAuth';
import Layout from 'components/Layout';
import Spinner from 'components/Spinner/Spinner';

export const UserContext = createContext({
  user: undefined,
  setUser: () => {},
});

export const AuthProvider = ({ children }) => {
  const [user, setUser] = useState();
  const [redirect, setRedirect] = useState();
  const [isLoading, setLoading] = useState(true);
  const router = useRouter();
  useEffect(() => {
    const initializeAuth = async () => {
      try {
        const data = getUser();
        setUser(data);
      } catch (e) {
        setUser();
        setRedirect(e.response.status === 403 ? '/access-denied' : '/login');
      }
      setLoading(false);
    };
    initializeAuth();
  }, []);
  if (isLoading || !router.isReady) {
    return (
      <Layout>
        <Spinner />
      </Layout>
    );
  }
  if (
    typeof window !== 'undefined' &&
    AUTH_WHITELIST.includes(router.pathname) === Boolean(user)
  ) {
    router.push(redirect || location);
    return <></>;
  }
  return (
    <UserContext.Provider
      value={{
        user,
        isLoading,
        setUser,
        redirect,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};

export function useAuth() {
  const context = useContext(UserContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
}
