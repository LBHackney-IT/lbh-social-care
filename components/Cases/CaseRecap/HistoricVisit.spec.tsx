import { render } from '@testing-library/react';

import { UserContext } from 'components/UserContext/UserContext';
import { mockedUser } from 'factories/users';
import * as casesAPI from 'utils/api/cases';
import HistoricNote from './HistoricNote';

import { mockedHistoricVisitNote } from 'fixtures/cases.fixtures';

jest.mock('components/Spinner/Spinner', () => () => 'MockedSpinner');

describe(`HistoricNote`, () => {
  it('should display historic case note', async () => {
    const props = {
      recordId: '123',
      is_historical: true,
      personId: 123,
    };
    jest.spyOn(casesAPI, 'useHistoricCaseVisit').mockImplementation(() => ({
      data: mockedHistoricVisitNote,
      isValidating: false,
      mutate: jest.fn(),
      revalidate: jest.fn(),
    }));
    const { asFragment } = render(
      <UserContext.Provider
        value={{
          user: mockedUser,
        }}
      >
        <HistoricNote {...props} />
      </UserContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
