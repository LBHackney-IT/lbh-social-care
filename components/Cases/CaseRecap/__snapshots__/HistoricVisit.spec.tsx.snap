// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistoricNote should display historic case note 1`] = `
<DocumentFragment>
  <span
    class="govuk-error-message"
  >
    <span
      class="govuk-visually-hidden"
    >
      Error:
    </span>
     Oops an error occurred
  </span>
</DocumentFragment>
`;
